Source/Core/InputCommon/ControllerEmu/ControlGroup/Cursor.cpp

```diff
@@ -31,89 +32,121 @@ Cursor::Cursor(const std::string& name_) : ControlGroup(name_, GroupType::Cursor
-  }
+  const double max_step = STEP_PER_SEC / 1000.0 * ms_since_update;
+  const double max_z_step = STEP_Z_PER_SEC / 1000.0 * ms_since_update;
 
-  // hide
-  const bool autohide = boolean_settings[1]->GetValue() && m_autohide_timer < 0;
-  if (controls[6]->control_ref->State() > 0.5 || autohide)
-  {
-    result.x = 10000;
-    result.y = 0;
-  }
-  else
-  {
-    // adjust cursor according to settings
-    if (adjusted)
-    {
-      xx *= (numeric_settings[1]->GetValue() * 2);
-      yy *= (numeric_settings[2]->GetValue() * 2);
-      yy += (numeric_settings[0]->GetValue() - 0.5);
-    }
+  // Apply deadzone to z:
+  const ControlState deadzone = numeric_settings[SETTING_DEADZONE]->GetValue();
+  z = std::copysign(std::max(0.0, std::abs(z) - deadzone) / (1.0 - deadzone), z);
+
+  // Smooth out z movement:
+  // FYI: Not using relative input for Z.
+  m_state.z += MathUtil::Clamp(z - m_state.z, -max_z_step, max_z_step);
 
-    // relative input
-    if (boolean_settings[0]->GetValue())
+  // Relative input:
+  if (boolean_settings[0]->GetValue())
+  {
+    // Recenter:
+    if (controls[7]->control_ref->State() > BUTTON_THRESHOLD)
     {
-      // deadzone to avoid the cursor slowly drifting
-      if (std::abs(xx) > deadzone)
-        m_state.x = MathUtil::Clamp(m_state.x + xx * SPEED_MULTIPLIER, -1.0, 1.0);
-      if (std::abs(yy) > deadzone)
-        m_state.y = MathUtil::Clamp(m_state.y + yy * SPEED_MULTIPLIER, -1.0, 1.0);
-
-      // recenter
-      if (controls[7]->control_ref->State() > 0.5)
-      {
-        m_state.x = 0.0;
-        m_state.y = 0.0;
-      }
+      m_state.x = 0.0;
+      m_state.y = 0.0;
     }
     else
     {
-      m_state.x = xx;
-      m_state.y = yy;
+      m_state.x = MathUtil::Clamp(m_state.x + input.x * max_step, -1.0, 1.0);
+      m_state.y = MathUtil::Clamp(m_state.y + input.y * max_step, -1.0, 1.0);
     }
+  }
+  // Absolute input:
+  else
+  {
+    m_state.x = input.x;
+    m_state.y = input.y;
+  }
+
+  StateData result = m_state;
+
+  // Adjust cursor according to settings:
+  result.x *= (numeric_settings[SETTING_WIDTH]->GetValue() * 2);
+  result.y *= (numeric_settings[SETTING_HEIGHT]->GetValue() * 2);
+  result.y += (numeric_settings[SETTING_CENTER]->GetValue() - 0.5);
+
+  const bool autohide = boolean_settings[1]->GetValue();
 
-    result.x = m_state.x;
-    result.y = m_state.y;
+  // Auto-hide timer:
+  // TODO: should Z movement reset this?
+  if (!autohide || std::abs(m_prev_result.x - result.x) > AUTO_HIDE_DEADZONE ||
+      std::abs(m_prev_result.y - result.y) > AUTO_HIDE_DEADZONE)
+  {
+    m_auto_hide_timer = AUTO_HIDE_MS;
+  }
+  else if (m_auto_hide_timer)
+  {
+    m_auto_hide_timer -= std::min<int>(ms_since_update, m_auto_hide_timer);
   }
 
-  m_prev_xx = xx;
-  m_prev_yy = yy;
+  m_prev_result = result;
+
+  // If auto-hide time is up or hide button is held:
+  if (!m_auto_hide_timer || controls[6]->control_ref->State() > BUTTON_THRESHOLD)
+  {
+    // TODO: Use NaN or something:
```

Comment:
Do you intend to take care of this TODO within the scope of this PR?

Reply:
Nope. :P It's just an ugly magic number that is handled elsewhere.